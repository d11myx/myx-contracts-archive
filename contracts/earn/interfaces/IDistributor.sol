// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

interface IDistributor {
    function updateRoot(bytes32 _merkleRoot) external;
    function claimForAccount(address account, address receiver, uint256 amount, bytes32[] calldata merkleProof) external;
}
