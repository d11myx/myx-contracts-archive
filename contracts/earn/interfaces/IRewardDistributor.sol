// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

interface IRewardDistributor {
    function updateRoot(bytes32 _merkleRoot, uint256 _amount) external;

    function claimForAccount(
        address account,
        address receiver,
        uint256 amount,
        bytes32[] calldata merkleProof
    ) external;
}
