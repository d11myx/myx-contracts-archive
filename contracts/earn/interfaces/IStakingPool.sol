// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

interface IStakingPool {
    function stakeForAccount(address funder, address account, uint256 amount) external;
    function unstakeForAccount(address account, address receiver, uint256 amount) external;
}
