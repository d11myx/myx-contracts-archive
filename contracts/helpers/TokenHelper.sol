// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
import "@openzeppelin/contracts/utils/math/SafeMath.sol";
import "../libraries/PrecisionUtils.sol";
import "../interfaces/IPool.sol";

library TokenHelper {
    using PrecisionUtils for uint256;
    using SafeMath for uint256;

    function convertIndexAmountToStable(
        IPool.Pair memory pair,
        int256 indexTokenAmount
    ) internal view returns (int256 amount) {
        uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
        uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

        uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
        uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

        amount = (indexTokenAmount * int256(indexTokenWad)) / int256(stableTokenWad);
    }

    // function convertIndexAmountToStableWithPrice(
    //     IPool.Pair memory pair,
    //     int256 indexTokenAmount,
    //     uint256 price
    // ) internal view returns (int256 amount) {
    //     uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
    //     uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

    //     uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
    //     uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

    //     amount =
    //         ((indexTokenAmount * int256(price) * int256(indexTokenWad)) /
    //             int256(PrecisionUtils.PRICE_PRECISION)) *
    //         int256(stableTokenWad);
    // }

    function convertStableAmountToIndex(
        IPool.Pair memory pair,
        int256 stableTokenAmount
    ) internal view returns (int256 amount) {
        uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
        uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

        uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
        uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

        amount = (stableTokenAmount * int256(stableTokenWad)) / int256(indexTokenWad);
    }
}
