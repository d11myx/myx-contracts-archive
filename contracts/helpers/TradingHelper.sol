// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import '../libraries/PrecisionUtils.sol';
import '../interfaces/IAddressesProvider.sol';
import '../interfaces/IOraclePriceFeed.sol';
import '../interfaces/IIndexPriceFeed.sol';
import '../interfaces/IPool.sol';

// import 'hardhat/console.sol';

library TradingHelper {
    using PrecisionUtils for uint256;

    function getValidPrice(
        IAddressesProvider addressesProvider,
        address token,
        IPool.TradingConfig memory tradingConfig
    ) internal view returns (uint256) {
        uint256 oraclePrice = IOraclePriceFeed(addressesProvider.priceOracle()).getPrice(token);

        uint256 indexPrice = IIndexPriceFeed(addressesProvider.indexPriceOracle()).getPrice(token);

        uint256 diffP = oraclePrice > indexPrice ? oraclePrice - indexPrice : indexPrice - oraclePrice;
        diffP = diffP.calculatePercentage(oraclePrice);

        require(diffP <= tradingConfig.maxPriceDeviationP, 'exceed max price deviation');
        return oraclePrice;
    }
}
