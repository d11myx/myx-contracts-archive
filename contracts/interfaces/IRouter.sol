// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import "../libraries/type/TradingTypes.sol";

interface IRouter {

    event UpdateTradingRouter(address oldAddress, address newAddress);

    function createIncreaseOrder(TradingTypes.IncreasePositionRequest memory request) external returns (uint256);

    function createDecreaseOrder(TradingTypes.DecreasePositionRequest memory request) external returns (uint256);

    function cancelIncreaseOrder(uint256 orderId, TradingTypes.TradeType tradeType) external;

    function cancelDecreaseOrder(uint256 orderId, TradingTypes.TradeType tradeType) external;

    function cancelAllPositionOrders(uint256 pairIndex, bool isLong) external;

    function cancelOrders(uint256 pairIndex, bool isLong, bool isIncrease) external;

    function createTpSl(TradingTypes.CreateTpSlRequest memory request) external returns (uint256 tpOrderId, uint256 slOrderId);

}
