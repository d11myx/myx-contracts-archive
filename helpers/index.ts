export * from './constants';
export * from './contract-deployments';
export * from './contract-getters';
export * from './deploy-ids';
export * from './env';
export * from './hardhat-config-helpers';
export * from './init-helper';
export * from './market-config-helper';
export * from './types';
export * from './utilities/signer';
export * from './utilities/tx';
export * from '../tasks/misc/print-deployments';

import { loadTasks } from './hardhat-config-helpers';

const TASK_FOLDERS = ['../tasks/misc'];

// Load all plugin tasks
loadTasks(TASK_FOLDERS);
