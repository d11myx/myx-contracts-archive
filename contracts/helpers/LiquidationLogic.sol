// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.20;

import "../libraries/Position.sol";
import "../interfaces/IPool.sol";
import "../interfaces/IOrderManager.sol";
import "../interfaces/IPositionManager.sol";
import "../interfaces/IAddressesProvider.sol";
import "../helpers/TradingHelper.sol";
import "../interfaces/IExecutor.sol";

library LiquidationLogic {
    using PrecisionUtils for uint256;
    using Position for Position.Info;

    uint256 public constant MAX_RATIO = 3e6;

    event ExecuteLiquidation(
        bytes32 positionKey,
        address account,
        uint256 pairIndex,
        bool isLong,
        uint256 collateral,
        uint256 sizeAmount,
        uint256 price
    );

    function liquidationPosition(
        IPool pool,
        IOrderManager orderManager,
        IPositionManager positionManager,
        IExecutor executor,
        IAddressesProvider addressesProvider,
        bytes32 positionKey,
        uint8 level,
        uint256 commissionRatio
    ) external {
        require(commissionRatio < MAX_RATIO, "max ratio");
        Position.Info memory position = positionManager.getPositionByKey(positionKey);
        if (position.positionAmount == 0) {
            return;
        }
        IPool.Pair memory pair = pool.getPair(position.pairIndex);
        IPool.TradingConfig memory tradingConfig = pool.getTradingConfig(position.pairIndex);

        uint256 price = TradingHelper.getValidPrice(
            addressesProvider,
            pair.indexToken,
            tradingConfig
        );

        int256 unrealizedPnl = position.getUnrealizedPnl(position.positionAmount, price);
        uint256 tradingFee = positionManager.getTradingFee(
            position.pairIndex,
            position.isLong,
            position.positionAmount
        );
        int256 fundingFee = positionManager.getFundingFee(
            position.account,
            position.pairIndex,
            position.isLong
        );
        int256 exposureAsset = int256(position.collateral) +
            unrealizedPnl -
            int256(tradingFee) +
            fundingFee;

        bool needLiquidate;
        if (exposureAsset <= 0) {
            needLiquidate = true;
        } else {
            uint256 riskRate = position
                .positionAmount
                .mulPrice(position.averagePrice)
                .mulPercentage(tradingConfig.maintainMarginRate)
                .calculatePercentage(uint256(exposureAsset));
            needLiquidate = riskRate >= PrecisionUtils.percentage();
        }
        if (!needLiquidate) {
            return;
        }

        // cancel all positionOrders
        orderManager.cancelAllPositionOrders(position.account, position.pairIndex, position.isLong);

        uint256 orderId = orderManager.createOrder(
            TradingTypes.CreateOrderRequest({
                account: position.account,
                pairIndex: position.pairIndex,
                tradeType: TradingTypes.TradeType.MARKET,
                collateral: 0,
                openPrice: price,
                isLong: position.isLong,
                sizeAmount: -int256(position.positionAmount),
                maxSlippage: 0,
                data: abi.encode(position.account)
            })
        );

        executor.executeDecreaseOrder(
            orderId,
            TradingTypes.TradeType.MARKET,
            level,
            commissionRatio
        );

        emit ExecuteLiquidation(
            positionKey,
            position.account,
            position.pairIndex,
            position.isLong,
            position.collateral,
            position.positionAmount,
            price
        );
    }
}
