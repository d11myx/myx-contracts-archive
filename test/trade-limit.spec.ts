import { newTestEnv, TestEnv } from './helpers/make-suite';
import { expect } from './shared/expect';
import { ethers } from 'hardhat';
import { increasePosition, mintAndApprove, updateBTCPrice } from './helpers/misc';
import { deployMockCallback, TradeType } from '../helpers';
import { TradingTypes } from '../types/contracts/interfaces/IRouter';
import snapshotGasCost from './shared/snapshotGasCost';

describe('Trade: Limit order cases', () => {
    const pairIndex = 0;
    let testEnv: TestEnv;

    before(async () => {
        testEnv = await newTestEnv();
        const {
            users: [depositor],
            usdt,
            btc,
            pool,
        } = testEnv;

        // add liquidity
        const indexAmount = ethers.utils.parseUnits('10000', 18);
        const stableAmount = ethers.utils.parseUnits('300000000', 18);
        let testCallBack = await deployMockCallback();
        const pair = await pool.getPair(pairIndex);

        await mintAndApprove(testEnv, btc, indexAmount, depositor, testCallBack.address);
        await mintAndApprove(testEnv, usdt, stableAmount, depositor, testCallBack.address);

        await snapshotGasCost(
            testCallBack
                .connect(depositor.signer)
                .addLiquidity(pool.address, pair.indexToken, pair.stableToken, indexAmount, stableAmount),
        );
    });

    describe('long > short', () => {
        it('long > short, increase order, open long position', async () => {
            const {
                keeper,
                users: [, trader],
                usdt,
                positionManager,
                orderManager,
                router,
                executor,
            } = testEnv;

            // update BTC price
            await updateBTCPrice(testEnv, '32000');

            const collateral = ethers.utils.parseUnits('20000', 18);
            const increaseAmount = ethers.utils.parseUnits('10', 18);
            const triggerPrice = ethers.utils.parseUnits('32000', 30);

            await mintAndApprove(testEnv, usdt, collateral, trader, router.address);
            const orderId = await increasePosition(
                testEnv,
                trader,
                pairIndex,
                collateral,
                triggerPrice,
                increaseAmount,
                TradeType.LIMIT,
                true,
            );

            const positionAft = await positionManager.getPosition(trader.address, pairIndex, true);
            console.log(`---positionAft: `, positionAft);

            await executor
                .connect(keeper.signer)
                .executeIncreaseLimitOrders([{ orderId: orderId, level: 0, commissionRatio: 0 }]);
            // expect(positionAft.positionAmount).to.be.eq(size);
        });

        it('long > short, increase order, open short position', async () => {});

        it('long > short, decrease order, open long position', async () => {});

        it('long > short, decrease order, open short position', async () => {});
    });

    describe('long < short', () => {
        //     before(async () => {
        //         const {
        //             users: [trader],
        //             usdt,
        //             positionManager,
        //             orderManager,
        //         } = testEnv;
        //
        //         const netExposureAmountBefore = await positionManager.getExposedPositions(pairIndex);
        //         expect(netExposureAmountBefore).to.be.eq(ethers.utils.parseUnits('1000', 18));
        //
        //         const collateral = ethers.utils.parseUnits('3000000', 18);
        //         const size = ethers.utils.parseUnits('2000', 18);
        //
        //         await mintAndApprove(testEnv, usdt, collateral, trader, orderManager.address);
        //         await increasePosition(testEnv, trader, pairIndex, collateral, size, TradeType.MARKET, false);
        //
        //         const netExposureAmountAfter = await positionManager.getExposedPositions(pairIndex);
        //         expect(netExposureAmountAfter).to.be.eq(netExposureAmountBefore.sub(size));
        //         expect(netExposureAmountAfter).to.be.lt(0);
        //     });

        it('long < short, increase order, open long position', async () => {});

        it('long < short, increase order, open short position', async () => {});

        it('long < short, decrease order, open long position', async () => {});

        it('long < short, decrease order, open short position', async () => {});
    });

    describe('long = short', () => {
        // before(async () => {
        //     const {
        //         users: [trader],
        //         positionManager,
        //     } = testEnv;
        //     const size = ethers.utils.parseUnits('1000', 18);
        //
        //     const netExposureAmountBefore = await positionManager.getExposedPositions(pairIndex);
        //     expect(netExposureAmountBefore).to.be.eq(BigNumber.from('-1000000000000000000000'));
        //
        //     await decreasePosition(testEnv, trader, pairIndex, BigNumber.from(0), size, TradeType.MARKET, false);
        //     const netExposureAmountAfter = await positionManager.getExposedPositions(pairIndex);
        //     expect(netExposureAmountAfter).to.be.eq(netExposureAmountBefore.add(size));
        //     expect(netExposureAmountAfter).to.be.eq(0);
        // });

        it('long = short, increase order, open long position', async () => {});

        it('long = short, increase order, open short position', async () => {});

        it('long = short, decrease order, open long position', async () => {});

        it('long = short, decrease order, open short position', async () => {});
    });
});
