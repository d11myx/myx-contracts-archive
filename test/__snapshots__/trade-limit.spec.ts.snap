// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trade: Limit order cases long > short long > short, increase order, open long position 1`] = `71517`;
